api/MetricsApi.ts

```
@@ -14,8 +14,20 @@ async function findAll(): Promise<MetricBare[]> {
   return (await fetchApi('GET', '/metrics')).metrics.map((apiData: ApiData) => MetricBare.fromApiData(apiData))
 }
 
+/**
+ * Find the metric by ID.
+ *
+ * Note: Be sure to handle any errors that may be thrown.
+ *
+ * @throws UnauthorizedError
+ */
+async function findById(metricId: number): Promise<MetricFull> {
+  return MetricFull.fromApiData(await fetchApi('GET', `/metrics/${metricId}`))
```

FWIW, this will fail spectacularly with an undefined/nonexistant metric id. However, that's out of scope for this PR. Maybe this weekend or next week, I'll dip my toes into some TS and see what I come up with :) if someone doesn't beat me to it.

There's two different ways to handle errors:

1. Throw an exception in `fetchApi()` and rely on error boundaries and try-catch-finally
2. Return the api object OR an error object.

I'm partial to first one because I think that errors are exceptional in this case and that it allows us to uniformly handle errors. I'm also super annoyed dealing with functions that may or may not return errors (:sigh: PHP)

---

I can temporarily hook this function up to the production API so I can write the error handling code now since the mock API is very limited now.

I am good with throwing an exception or returning an error object. I think it is best not to handle the error within the call because that couples the UI too tightly to the API code. (Not implying you were suggesting to handle the error here in the API layer.)

The Swagger Yaml says a 404 should be returned. I think throwing an `NotFound` error would be good. The UI that calls this API function can then handle it appropriately.

---

@withinboredom @yanirs I temporarily hooked up the API code to the production API. A status of 404 is properly returned along with a JSON body matching the Error schema:
```
{
  "code": "metric_not_found",
  "message": "The metric was not found",
  "data": {
    "status": 404
  }
}
```

FYI, here is the list of status codes the API is documented as returning:

200 - Success
201 - Created
400 - BadRequest
401 - Unauthorized
403 - Forbidden
404 - NotFound
409 - Conflict
501 - NotImplemented

We can handle this several ways.
1) Assuming all works as expected, the API always returns a body with JSON regardless of the response status.
2) In `api/utils.ts`, check the response status, parse the body as JSON, if response is anything but 200 or 201, then parse body should be an error object. This can be passed to an ApiError class we create and it can be thrown.
3) Same as #2 but return the ApiError instance.
4) Similar to #2 but have separate error instances for the different error statuses. For example,  a NotFoundError for 404s, UnauthorizedError (this error already exists) for 401s, etc.

Before making a decision on how to handle it in the API layer, it might be useful to think about how the UI layer will handle an API error. Here is what I am imagining:

If the error is a 401, the UI redirects or prompts the user to re-authenticate.
If the error is a 403 or a 404, then give the user an appropriate UI message.
I think the other error states (400, 409, 501, and any other error condition the server may response with) can be handled as such:
  a) Display an appropriate UI message. Can just be generic since the details are usually meaningless to the user.
  b) Console log the error details.

Also, to keep the current code clean and simple I think throwing errors is better.
Currently, we have something like:
```
  return (await fetchApi('GET', '/metrics')).metrics.map((apiData: ApiData) => MetricBare.fromApiData(apiData))
```
This works perfectly in a non-error situation. But if we don't throw an error in `api/utils.ts`, then each one of the functions like the one above will need special, repeated error handling. I think it would be clean if the error trickles up to the UI layer for appropriate handling. Then these functions can concentrate on the success state.

Thoughts? P2 this?

---

You can also let it bubble up to an error boundary, which imho is probably cleaner and more reusable. When looking at the code, you need to only consider the “happy path”

```jsx
<NotFound show={<>Unable to locate metric</>}>
  <MetricThing forId={42} />
</ NotFound>
```

---

The error boundary and throwing an `ApiError` that basically wraps the API error information can help centralize the UI error handling logic.

---

+1 to centralising the logic for throwing and handling API errors by type, as suggested. Definitely out of scope for this PR, so can you please open an issue? It feels like a lower priority to me than getting the basic views out &ndash; it's more of a polishing step.